### PR TITLE
[ogr2ogr] Fix promote to multipart logic and add makevalid option

### LINKED ISSUE #58392: Improve handling of  import options  in Export to PostgreSQL ogr2ogr
### Feature description

Currently when importing a dataset into PostgreSQL using the ogr2ogr algorithm, if you set the following:
Screenshot 2024-08-14 at 15 12 15

By default, the other option for converting to multi is also set i.e
Screenshot 2024-08-14 at 15 13 25

This will lead to a command line being generated like

```
ogr2ogr -progress --config PG_USE_COPY YES -f PostgreSQL "PG:dbname='gis' host=localhost port=5432 user='docker' password='docker' sslmode=allow active_schema=public" -lco DIM=2 /Applications/QGIS.app/Contents/Resources/resources/data/world_map.gpkg countries -overwrite -nlt POLYGON -lco GEOMETRY_NAME=geom -lco FID=id -nln public.countries -nlt PROMOTE_TO_MULTI
```

This leads to having
```
 -nlt POLYGON  -nlt PROMOTE_TO_MULTI

```
 
 

### Additional context

You can always untick the `Promote to MultiPart` so your algorithm works properly. I think it would make sense to implement the following logic

# Solution A
* If the geometry type is unset then we default to have Promote to Multipart active by default

# Solution B

* If the geometry type option is selected then deactivate `Promote to Multipart`

# Solution C
* by default have the `Promote to Multipart` deactivated

**Note** There are other options like overwriting the existing table and appending to an existing table that should be mutually exclusive

### PR DESCRIPTION
**Pull request by NyakudyaA**
Fixes https://github.com/qgis/QGIS/issues/58392

- [x] Add an option for makevalid, off by default
- [x] Raise an exception if Promote and  output geometry type are both selected
- [x] Raise an exception if Append and Overwrite are both selected

Screenshot 2024-08-20 at 10 46 59

### CODE REVIEW

**Comment by agiudiceandrea** [comment]
Maybe it could be better to inform the user or raise an exception about the fact that they has selected the two incompatible options GTYPE and PROMOTETOMULTI (also when the algorithm is called by a Python script or it is used in a module).
It looks like the CONVERT_TO_LINEAR and CONVERT_TO_CURVE options are missing for the `-nlt` parameter, anyway they could be added as additional command line parameter.
Isn't the "Export to PostgreSQL (new connection)" (`gdal:importvectorintopostgisdatabasenewconnection`) also affected?
What about the "overwriting the existing table and appending to an existing table that should be mutually exclusive" reported in the same issue report?

**Comment by NyakudyaA** [comment]
> Maybe it could be better to inform the user or raise an exception about the fact that they has selected the two incompatible options GTYPE and PROMOTETOMULTI (also when the algorithm is called by a Python script or it is used in a module). It looks like the CONVERT_TO_LINEAR and CONVERT_TO_CURVE options are missing for the `-nlt` parameter, anyway they could be added as additional command line parameter. Isn't the "Export to PostgreSQL (new connection)" (`gdal:importvectorintopostgisdatabasenewconnection`) also affected? What about the "overwriting the existing table and appending to an existing table that should be mutually exclusive" reported in the same issue report?

I updated the other script and added the missing import options for GTYPE. Now we get the following
Screenshot 2024-08-20 at 13 37 13

**Comment by agiudiceandrea** [comment]
Thanks!
According to https://gdal.org/programs/ogr2ogr.html#cmdoption-ogr2ogr-nlt it looks like that "-nlt CONVERT_TO_LINEAR and -nlt PROMOTE_TO_MULTI can be used simultaneously."

**Comment by agiudiceandrea** [comment]
@rouault it looks like the `-overwrite` and the `-append` options for ogr2ogr has been declared as mutually exclusive only since GDAL/OGR 3.9.0, rising the `Argument '-overwrite' not allowed with '-append'` error, while until GDAL/OGR 3.8.5 they could be both present at the same time without rising an error (in such case it seems to me only the last one in the command was taken in consideration). Is / was it the intended behaviour? Do you think the algorithm should allow `-append -overwrite` when used with GDAL/OGR < 3.9.0 and only rise an exception for GDAL/OGR >= 3.9.0, or should it always rise an exception regardless of the GDAL/OGR version used?

**Comment by rouault** [comment]
-overwrite and -append have always been mutually exclusive. It is just that GDAL 3.9 starts making it an error. In previous versions, the last specified argument would win over the previously secified. It makes sense for QGIS to treat them as mutually exlusive for any GDAL version

**Comment by NyakudyaA** [comment]
@agiudiceandrea, should we also handle the `-overwrite` and `-addfields` as mutually exclusive? Should we also add some validation logic for the `Additional creation options`, I assume if future gdal versions error in -overwrite and -append, then it should be handled or do we let users handle this since this option is assumed to be useful for advanced users

**Comment by agiudiceandrea** [comment]
> should we also handle the `-overwrite` and `-addfields` as mutually exclusive?

It look like they are not actually enforced as mutually exclusive in ogr2ogr, but probably they should be.

> Should we also add some validation logic for the `Additional creation options`

AFAIK there is no validation in other algorithms for the additional creation options, but I've not actually checked.

**Comment by NyakudyaA** [comment]
Not stale

**Comment by NyakudyaA** [comment]
Not stale

**Comment by NyakudyaA** [comment]
Not stale

File: python/plugins/processing/algs/gdal/ogr2ogrtopostgislist.py [hunk 1]
```diff
@@ -316,7 +330,17 @@ def getConsoleCommands(self, parameters, context, feedback, executing=True):
         if len(gt) > 0:
             arguments.append('-gt')
             arguments.append(gt)
-        if promotetomulti:
+        if make_valid:
+            arguments.append('-makevalid')
+        if promotetomulti and len(self.GEOMTYPE[self.parameterAsEnum(parameters, self.GTYPE, context)]) > 0:
```
Comment (nyalldawson) [inline python/plugins/processing/algs/gdal/ogr2ogrtopostgislist.py:335]:
```suggestion
        if promotetomulti and self.GEOMTYPE[self.parameterAsEnum(parameters, self.GTYPE, context)]:
```

File: python/plugins/processing/algs/gdal/ogr2ogrtopostgislist.py [hunk 2]
```diff
@@ -316,7 +330,17 @@ def getConsoleCommands(self, parameters, context, feedback, executing=True):
...
+            if self.GEOMTYPE[self.parameterAsEnum(parameters, self.GTYPE, context)] == 'CONVERT_TO_LINEAR':
+                arguments.append('-nlt PROMOTE_TO_MULTI')
+            else:
+                raise QgsProcessingException(
+                    self.tr(
+                        'Only one of Promote to Multipart or Output Geometry Type (excluding Convert to Linear) can be enabled.'))
```
Comment (nyalldawson) [inline python/plugins/processing/algs/gdal/ogr2ogrtopostgislist.py:341]:
```suggestion
                        'Only one of "Promote to Multipart" or "Output geometry type" (excluding Convert to Linear) can be enabled.'))
```

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
@agiudiceandrea was there more here you wanted to see?

**Comment by agiudiceandrea** [comment]
Looks good to me.

**Pull request merged**